scripts/run_nodes.sh

```diff
@@ -1,123 +1,289 @@
 
-  if $BOOTSTRAP ; then
-    ./target/release/aleph-node purge-chain --base-path $BASE_PATH/$account_id --chain $BASE_PATH/chainspec.json -y
+  info "Generating keystores for ${RPC_NODES} RPC nodes"
+  for i in $(seq 0 "$(( RPC_NODES - 1 ))"); do
+    rpc_node_account_id="${rpc_node_account_ids[$i]}"
+    "${ALEPH_NODE}" bootstrap-node --base-path "${BASE_PATH}/${rpc_node_account_id}" --account-id "${rpc_node_account_id}" --chain-type local > /dev/null
+  done
+
+  if [[ "${DONT_REMOVE_ABFT_BACKUPS}" == "true" ]]; then
+    non_empty_backups=$(find "${BASE_PATH}" -type d -name "*backup-stash*" -not -empty)
+    if [[ -n "${non_empty_backups}" ]]; then
+      warning "Found following non-empty ABFT backups in base path:"
+      warning "${non_empty_backups}"
+      warning "In 99% you want them to be removed when bootstraping chain"
+      warning "Re-run this script with without flag --dont-remove-abtf-backups if you want them to be removed."
+    fi
   fi
-  ./target/release/aleph-node \
-    $validator \
-    --chain $BASE_PATH/chainspec.json \
-    --base-path $BASE_PATH/$account_id \
-    --name $auth \
-    --rpc-port $((9943 + i)) \
-    --port $((30334 + i)) \
-    --bootnodes $bootnodes \
-    --node-key-file $BASE_PATH/$account_id/p2p_secret \
-    --backup-path $BASE_PATH/$account_id/backup-stash \
-    --execution Native \
-    --rpc-cors=all \
-    --no-mdns \
-    --public-validator-addresses 127.0.0.1:${validator_port} \
-    --validator-port ${validator_port} \
-    --detailed-log-output \
-    -laleph-party=debug \
-    -laleph-network=debug \
-    -lnetwork-clique=debug \
-    -laleph-finality=debug \
-    -laleph-justification=debug \
-    -laleph-data-store=debug \
-    -laleph-updater=debug \
-    -laleph-metrics=debug \
-    2> $auth.log > /dev/null & \
-}
+fi
+
+info "Creating bootnode p2p multiaddress."
+p2p_key_public=$("${ALEPH_NODE}" key inspect-node-key --file "${BASE_PATH}/${rpc_node_account_ids[0]}/p2p_secret")
+bootnode_multiaddress="/dns4/localhost/tcp/$((NODE_P2P_PORT_RANGE_START))/p2p/${p2p_key_public}"
+info "Bootnode p2p multiaddress is ${bootnode_multiaddress}"
+
+if [[ -z "${DONT_DELETE_DB}" ]] ; then
+  info "Removing database for all nodes (aka purging chain)."
+  for i in $(seq 0 "$(( RPC_NODES - 1 ))"); do
+    rpc_node_account_id=${rpc_node_account_ids[$i]}
+    "${ALEPH_NODE}" purge-chain --base-path "${BASE_PATH}/${rpc_node_account_id}" --chain "${BASE_PATH}/chainspec.json" -y > /dev/null 2>&1
+  done
+  for i in $(seq 0 "$(( VALIDATORS - 1 ))"); do
+    validator_account_id="${validator_account_ids[$i]}"
+    "${ALEPH_NODE}" purge-chain --base-path "${BASE_PATH}/${validator_account_id}" --chain "${BASE_PATH}/chainspec.json" -y > /dev/null 2>&1
+  done
+fi
 
-for i in $(seq 0 "$(( N_VALIDATORS + N_NON_VALIDATORS - 1 ))"); do
-  run_node "$i" true
+if [[ -z "${DONT_REMOVE_ABFT_BACKUPS}" ]]; then
+  info "Removing AlephBFT backups."
+  find "${BASE_PATH}" -type d -name "backup-stash" | xargs rm -rf
```

Comment:
Can we use a loop here instead (also for sake of consistency)? It has potential for removing too many (e.g. someones backup of a backup) or too few folders.

Reply:
Fixed

Reply:
Do we still need to use that `find` here then? If it is somehow unnecessary difficult without it then leave it. Since you have account-ids you could build those paths by yourself? IMHO it is nicer (and sometimes safer) when a clean-up script "knows" explicitly what needs to be removed, instead of searching/guessing.

Reply:
Good point, fixed